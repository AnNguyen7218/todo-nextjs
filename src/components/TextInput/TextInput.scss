@import 'config.fonts';
@import 'config.gutters';
@import 'config.colors';
@import 'config.media';

$input-padding-vertical-sm: gutter(10);
$input-padding-vertical-xl: gutter(12);
$input-padding-horizontal-sm: gutter(16);
$input-padding-horizontal-xl: gutter(16);

$password-icon-size-sm: gutter(16);
$password-icon-size-xl: gutter(24);

$input-height-sm: 4rem;
$input-height-xl: 4.8rem;

.input-wrapper {
  width: 100%;
  min-height: 8rem;

  @include break(xl) {
    min-height: 8.8rem;
  }
}

.label {
  display: block;
  @include small-medium;
}

.input-container {
  position: relative;
  width: 100%;
}

.input {
  @include body;
  width: 100%;
  padding: $input-padding-vertical-sm $input-padding-horizontal-sm;
  border: 0;
  background-color: $light-gray;
  height: $input-height-sm;

  @include break(xl) {
    padding: $input-padding-vertical-xl $input-padding-horizontal-xl;
    height: $input-height-xl;
  }

  &-password {
    padding-right: calc(
      #{$input-padding-horizontal-sm} + #{$password-icon-size-sm} + 0.5rem
    );
    @include break(xl) {
      padding-right: calc(
        #{$input-padding-horizontal-xl} + #{$password-icon-size-xl} + 0.5rem
      );
    }
  }

  & ~ .focus-border {
    position: absolute;
    bottom: 0;
    left: 50%;
    width: 0;
    height: 0.2rem;
    background-color: $purple;
    transition: 0.4s;

    &--error {
      background-color: $red;
      width: 100%;
      left: 0;
    }

    &--active {
      background-color: $green;
      width: 100%;
      left: 0;
      transition: 0.4s;
    }
  }

  &:focus ~ .focus-border {
    width: 100%;
    transition: 0.4s;
    left: 0;
  }
}

.message {
  @include small;
  color: $gray;

  &--error {
    color: $red;
  }

  &--active {
    color: $green;
  }
}

.password-icon {
  position: absolute;
  top: calc((#{$input-height-sm} - #{$password-icon-size-sm}) / 2);
  right: $input-padding-horizontal-sm;
  width: $password-icon-size-sm;

  @include break(xl) {
    top: calc((#{$input-height-xl} - #{$password-icon-size-xl}) / 2);
    right: $input-padding-horizontal-xl;
    width: $password-icon-size-xl;
  }
}
