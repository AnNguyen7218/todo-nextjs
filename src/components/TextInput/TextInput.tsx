import React, { useState } from 'react';
import cx from 'classnames';
import PasswordShow from 'assets/icons/ic-password-show.svg';
import PasswordHide from 'assets/icons/ic-password-hide.svg';
import css from './TextInput.scss';

type Props = React.DetailedHTMLProps<
  React.InputHTMLAttributes<HTMLInputElement>,
  HTMLInputElement
> & {
  onChange?: ((event: React.ChangeEvent<HTMLInputElement>) => void) | undefined;
  value: string;
  isError?: boolean;
  placeholder?: string;
  className?: string;
  containerClassName?: string;
  type: string;
  label?: string;
  hint?: string;
  errorMessage?: string;
  labelClassName?: string;
  isActive?: boolean;
  showPasswordToggleIcon?: boolean;
};

const TextInput = React.forwardRef<HTMLInputElement, Props>(
  (
    {
      containerClassName,
      onChange,
      value,
      isError,
      placeholder,
      labelClassName,
      className,
      type,
      isActive,
      showPasswordToggleIcon,
      label = '',
      hint = '',
      errorMessage = '',
      ...props
    },
    ref
  ) => {
    const [dynamicType, setDynamicType] = useState(type);
    const [isPasswordShow, setIsPasswordShow] = useState(false);

    const inputProps = {
      type: dynamicType,
      onChange,
      value,
      placeholder,
      ...props,
    };

    const PasswordIcon = isPasswordShow ? PasswordShow : PasswordHide;

    const onPasswordIconClick = () => {
      setDynamicType(isPasswordShow ? 'password' : 'text');
      setIsPasswordShow(!isPasswordShow);
    };

    const message = isError ? errorMessage : hint;
    const isActiveInput = !isError && isActive;

    return (
      <div className={cx(css['input-wrapper'], containerClassName)}>
        {label && (
          <label className={cx(css['label'], labelClassName)} htmlFor={label}>
            {label}
          </label>
        )}
        <div className={css['input-container']}>
          <input
            {...inputProps}
            className={cx(
              css['input'],
              { [css['input-password']]: showPasswordToggleIcon },
              className
            )}
            id={label}
            ref={ref}
          />
          <span
            className={cx(css['focus-border'], {
              [css['focus-border--error']]: isError,
              [css['focus-border--active']]: isActiveInput,
            })}
          />
          {showPasswordToggleIcon && (
            <PasswordIcon
              data-testid="password-icon"
              onClick={onPasswordIconClick}
              className={css['password-icon']}
            />
          )}
        </div>
        {message && (
          <p
            className={cx(css['message'], {
              [css['message--error']]: isError,
              [css['message--active']]: isActiveInput,
            })}
          >
            {message}
          </p>
        )}
      </div>
    );
  }
);

export default TextInput;
