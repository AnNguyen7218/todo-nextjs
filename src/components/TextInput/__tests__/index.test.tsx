import React from 'react';
import { render, cleanup, screen, fireEvent } from '@testing-library/react';
import TextInput from '../TextInput';

afterEach(cleanup);

describe('TextInput Component', () => {
  test('TextInput should match snapshot', () => {
    let value = 'value';
    const onChange = (event: React.ChangeEvent<HTMLInputElement>) => {
      value = event.target.value;
    };

    const { container } = render(
      <>
        <TextInput
          label="text input"
          type="text"
          hint="hint"
          value={value}
          placeholder="placeholder"
          onChange={onChange}
        />
        <TextInput
          label="email input"
          type="email"
          hint="hint"
          value={value}
          placeholder="placeholder"
          required
          onChange={onChange}
        />
        <TextInput
          label="text input"
          type="text"
          hint="hint"
          value={value}
          placeholder="placeholder"
          isError
          errorMessage="Error message"
          onChange={onChange}
        />
        <TextInput
          label="text input"
          type="text"
          hint="hint"
          value={value}
          placeholder="placeholder"
          isError
          errorMessage="Error message"
          isActive
          onChange={onChange}
        />
        <TextInput
          label="text input"
          type="text"
          hint="hint"
          value={value}
          placeholder="placeholder"
          isActive
          onChange={onChange}
        />
        <TextInput
          label="password input"
          type="password"
          hint="hint"
          value={value}
          placeholder="placeholder"
          onChange={onChange}
        />

        <TextInput
          label="password input"
          type="password"
          hint="hint"
          value={value}
          placeholder="placeholder"
          onChange={onChange}
          showPasswordToggleIcon
        />
      </>
    );
    expect(container).toMatchSnapshot();
  });

  test('password should show/hide on click', async () => {
    let value = 'password';
    const onChange = (event: React.ChangeEvent<HTMLInputElement>) => {
      value = event.target.value;
    };

    const { asFragment } = render(
      <TextInput
        label="password input"
        type="password"
        hint="hint"
        placeholder="placeholder"
        value={value}
        onChange={onChange}
        showPasswordToggleIcon
      />
    );

    expect(asFragment()).toMatchSnapshot();

    const passwordIcon = screen.queryByTestId('password-icon');
    expect(passwordIcon).not.toBeNull();
    passwordIcon && fireEvent.click(passwordIcon);
    expect(asFragment()).toMatchSnapshot();
  });
});
