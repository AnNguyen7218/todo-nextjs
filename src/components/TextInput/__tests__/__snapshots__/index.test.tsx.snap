// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput Component TextInput should match snapshot 1`] = `
<div>
  <div
    class="input-wrapper"
  >
    <label
      class="label"
      for="text input"
    >
      text input
    </label>
    <div
      class="input-container"
    >
      <input
        class="input"
        id="text input"
        placeholder="placeholder"
        type="text"
        value="value"
      />
      <span
        class="focus-border"
      />
    </div>
    <p
      class="message"
    >
      hint
    </p>
  </div>
  <div
    class="input-wrapper"
  >
    <label
      class="label"
      for="email input"
    >
      email input
    </label>
    <div
      class="input-container"
    >
      <input
        class="input"
        id="email input"
        placeholder="placeholder"
        required=""
        type="email"
        value="value"
      />
      <span
        class="focus-border"
      />
    </div>
    <p
      class="message"
    >
      hint
    </p>
  </div>
  <div
    class="input-wrapper"
  >
    <label
      class="label"
      for="text input"
    >
      text input
    </label>
    <div
      class="input-container"
    >
      <input
        class="input"
        id="text input"
        placeholder="placeholder"
        type="text"
        value="value"
      />
      <span
        class="focus-border focus-border--error"
      />
    </div>
    <p
      class="message message--error"
    >
      Error message
    </p>
  </div>
  <div
    class="input-wrapper"
  >
    <label
      class="label"
      for="text input"
    >
      text input
    </label>
    <div
      class="input-container"
    >
      <input
        class="input"
        id="text input"
        placeholder="placeholder"
        type="text"
        value="value"
      />
      <span
        class="focus-border focus-border--error"
      />
    </div>
    <p
      class="message message--error"
    >
      Error message
    </p>
  </div>
  <div
    class="input-wrapper"
  >
    <label
      class="label"
      for="text input"
    >
      text input
    </label>
    <div
      class="input-container"
    >
      <input
        class="input"
        id="text input"
        placeholder="placeholder"
        type="text"
        value="value"
      />
      <span
        class="focus-border focus-border--active"
      />
    </div>
    <p
      class="message message--active"
    >
      hint
    </p>
  </div>
  <div
    class="input-wrapper"
  >
    <label
      class="label"
      for="password input"
    >
      password input
    </label>
    <div
      class="input-container"
    >
      <input
        class="input"
        id="password input"
        placeholder="placeholder"
        type="password"
        value="value"
      />
      <span
        class="focus-border"
      />
    </div>
    <p
      class="message"
    >
      hint
    </p>
  </div>
  <div
    class="input-wrapper"
  >
    <label
      class="label"
      for="password input"
    >
      password input
    </label>
    <div
      class="input-container"
    >
      <input
        class="input input-password"
        id="password input"
        placeholder="placeholder"
        type="password"
        value="value"
      />
      <span
        class="focus-border"
      />
      <ic-password-hide.svg
        classname="password-icon"
        data-testid="password-icon"
      />
    </div>
    <p
      class="message"
    >
      hint
    </p>
  </div>
</div>
`;

exports[`TextInput Component password should show/hide on click 1`] = `
<DocumentFragment>
  <div
    class="input-wrapper"
  >
    <label
      class="label"
      for="password input"
    >
      password input
    </label>
    <div
      class="input-container"
    >
      <input
        class="input input-password"
        id="password input"
        placeholder="placeholder"
        type="password"
        value="password"
      />
      <span
        class="focus-border"
      />
      <ic-password-hide.svg
        classname="password-icon"
        data-testid="password-icon"
      />
    </div>
    <p
      class="message"
    >
      hint
    </p>
  </div>
</DocumentFragment>
`;

exports[`TextInput Component password should show/hide on click 2`] = `
<DocumentFragment>
  <div
    class="input-wrapper"
  >
    <label
      class="label"
      for="password input"
    >
      password input
    </label>
    <div
      class="input-container"
    >
      <input
        class="input input-password"
        id="password input"
        placeholder="placeholder"
        type="text"
        value="password"
      />
      <span
        class="focus-border"
      />
      <ic-password-show.svg
        classname="password-icon"
        data-testid="password-icon"
      />
    </div>
    <p
      class="message"
    >
      hint
    </p>
  </div>
</DocumentFragment>
`;
